File: Interviews/GFG/Daily-Problems/2022/October/Day-27/Print-leaf-nodes-from-preorder-traversal-of-BST.cpp
class Solution {
public:
    vector<int> leafNodes (int arr[], int N) {
        vector<int> st, ans;
        for (int i = 0, j = 1; j < N; i++, j++) {
            if (arr[j] < arr[i]) {
                st.push_back(arr[i]);
            } else {
                bool leaf = 0;
                while (st.size() and arr[j] > st.back()) {
                    st.pop_back();
                    leaf = 1;
                }

                if (leaf) ans.push_back(arr[i]);
            }
        }

        ans.push_back(arr[N - 1]);
        return ans;
    }
};

File: Interviews/GFG/random-problems/Medium/Print-leaf-nodes-from-preorder-traversal-of-BST/Print-leaf-nodes-from-preorder-traversal-of-BST.py
class Solution:
	def leafNodes (self, arr, N):
		st, ans, i, j = [], [], 0, 1
        while j < N:
            if arr[j] < arr[i]:
                st.append(arr[i])
            else:
                leaf = 0
                while len(st) and arr[j] > st[-1]:
                    st.pop()
                    leaf = 1
                if leaf: ans.append(arr[i])
            i, j = i + 1, j + 1
        
        ans.append(arr[-1])
        return ans